src/scanpy/preprocessing/_pca/_dask_sparse.py

```diff
@@ -0,0 +1,183 @@
+    components_: NDArray[np.floating]
+    explained_variance_: NDArray[np.floating]
+    explained_variance_ratio_: NDArray[np.floating]
+    noise_variance_: NDArray[np.floating]
+
+    def fit(self, x: DaskArray) -> PCASparseFit: ...
+    def transform(self, X: DaskArray) -> DaskArray: ...
+
+
+@dataclass
+class PCASparseDask:
+    n_components: int | None = None
+
+    def fit(self, x: DaskArray) -> PCASparseFit:
+        # this method makes `self` into the fitted version
+        self = cast(PCASparseFit, self)
+        assert isinstance(x.shape, tuple)
+        self.n_components_ = (
+            min(x.shape[:2]) if self.n_components is None else self.n_components
+        )
+        self.n_samples_ = x.shape[0]
+        self.n_features_in_ = x.shape[1] if x.ndim == 2 else 1
+        self.dtype_ = x.dtype
+        covariance, self.mean_ = _cov_sparse_dask(x)
+        self.explained_variance_, self.components_ = scipy.linalg.eigh(
+            covariance, lower=False
+        )
+
+        # Arrange eigenvectors and eigenvalues in descending order
+        self.explained_variance_ = self.explained_variance_[::-1]
+        self.components_ = np.flip(self.components_, axis=1)
+        self.components_ = self.components_.T[: self.n_components_, :]
+
+        self.explained_variance_ratio_ = self.explained_variance_ / np.sum(
+            self.explained_variance_
+        )
+        if self.n_components_ < min(self.n_samples_, self.n_features_in_):
+            self.noise_variance_ = self.explained_variance_[self.n_components_ :].mean()
+        else:
+            self.noise_variance_ = np.array([0.0])
+        self.explained_variance_ = self.explained_variance_[: self.n_components_]
+
+        self.explained_variance_ratio_ = self.explained_variance_ratio_[
+            : self.n_components_
+        ]
+        return self
+
+    def transform(self: PCASparseFit, x: DaskArray) -> DaskArray:
+        if TYPE_CHECKING:
+            import dask.array.core as da
+        else:
+            import dask.array as da
+
+        def transform_block(
+            x_part: CSMatrix,
+            mean_: NDArray[np.floating],
+            components_: NDArray[np.floating],
+        ):
+            pre_mean = mean_ @ components_.T
+            mean_impact = np.ones((x_part.shape[0], 1)) @ pre_mean.reshape(1, -1)
+            return (x_part @ components_.T) - mean_impact
+
+        return da.map_blocks(
+            transform_block,
+            x,
+            mean_=self.mean_,
+            components_=self.components_,
+            chunks=(x.chunks[0], self.n_components_),
+            meta=np.zeros([0], dtype=x.dtype),
+            dtype=x.dtype,
+        )
+
+    def fit_transform(self, x: DaskArray, y: DaskArray | None = None) -> DaskArray:
+        if y is None:
+            y = x
+        return self.fit(x).transform(y)
+
+
+@overload
+def _cov_sparse_dask(
+    x: DaskArray, *, return_gram: Literal[False] = False, dtype: DTypeLike | None = None
```

Comment:
We never use `return_gram` so let's remove

Reply:
we do in the test I added.

Reply:
What was the resolution here?

Reply:
I think testing intermediate results is useful because it makes it easier to debug when something breaks after the “checkpoint”.

If we want to change the internal structure, it’s easy to remove or adapt that part of the test code.